File: tests/golden/01-basic-types.ts
/**
 * 測試 01: 基本型別與變數宣告
 */

// 基本型別
const strr: string = "hello";
const numm: number = 42;
const boool: boolean = true;
const nothing: void = undefined;

// 型別推斷
let inferredString = "world";
let inferredNumber = 3.14;
let inferredBoolean = false;

// any 與 unknown
let anyValue: any = 42;
anyValue = "string";
anyValue = true;

let unknownValue: unknown = 42;
// unknownValue.toString(); // 應該要有型別檢查

// Arrays
const numberss: number[] = [1, 2, 3, 4, 5];
const strings: Array<string> = ["a", "b", "c"];

// Tuples
const tuple: [string, number] = ["age", 30];
const tuple3: [string, number, boolean] = ["test", 1, true];

// 可選與預設值
function greet(name: string, age?: number, title: string = "Mr."): string {
  if (age) {
    return `${title} ${name}, age ${age}`;
  }
  return `${title} ${name}`;
}

export { strr, numm, boool, greet };

File: tests/golden/expected/01-basic-types.go
package main

import "fmt"

var Strr string = "hello"
var Numm int = 42
var Boool bool = true
var nothing interface{} = nil

var inferredString = "world"
var inferredNumber = 3.14
var inferredBoolean = false
var anyValue = 42

var unknownValue interface{} = 42

var numberss = []int{1, 2, 3, 4, 5}
var strings = []string{"a", "b", "c"}

type Tuple2_string_int struct {
	Item0 string
	Item1 int
}

var tuple = Tuple2_string_int{"age", 30}

type Tuple3_string_int_bool struct {
	Item0 string
	Item1 int
	Item2 bool
}

var tuple3 = Tuple3_string_int_bool{"test", 1, true}

func Greet(name string, age *int, title string) string {
	if title == "" {
		title = "Mr."
	}
	if age != nil {
		return fmt.Sprintf("%s %s, age %v", title, name, *age)
	}
	return fmt.Sprintf("%s %s", title, name)
}
